File: src/shell/builtin/mkdir.zig
const Mkdir = @This();

opts: Opts = .{},
state: union(enum) {
    idle,
    exec: struct {
        started: bool = false,
        tasks_count: usize = 0,
        tasks_done: usize = 0,
        output_waiting: u16 = 0,
        output_done: u16 = 0,
        args: []const [*:0]const u8,
        err: ?jsc.SystemError = null,
    },
    waiting_write_err,
    done,
} = .idle,

pub fn onIOWriterChunk(this: *Mkdir, _: usize, e: ?jsc.SystemError) Yield {
    if (e) |err| err.deref();

    switch (this.state) {
        .waiting_write_err => return this.bltn().done(1),
        .exec => {
            this.state.exec.output_done += 1;
        },
        .idle, .done => @panic("Invalid state"),
    }

    return this.next();
}

pub fn writeFailingError(this: *Mkdir, buf: []const u8, exit_code: ExitCode) Yield {
    if (this.bltn().stderr.needsIO()) |safeguard| {
        this.state = .waiting_write_err;
        return this.bltn().stderr.enqueue(this, buf, safeguard);
    }

    _ = this.bltn().writeNoIO(.stderr, buf);
    // if (this.bltn().writeNoIO(.stderr, buf).asErr()) |e| {
    //     return .{ .err = e };
    // }

    return this.bltn().done(exit_code);
}

pub fn start(this: *Mkdir) Yield {
    const filepath_args = switch (this.opts.parse(this.bltn().argsSlice())) {
        .ok => |filepath_args| filepath_args,
        .err => |e| {
            const buf = switch (e) {
                .illegal_option => |opt_str| this.bltn().fmtErrorArena(.mkdir, "illegal option -- {s}\n", .{opt_str}),
                .show_usage => Builtin.Kind.mkdir.usageString(),
                .unsupported => |unsupported| this.bltn().fmtErrorArena(.mkdir, "unsupported option, please open a GitHub issue -- {s}\n", .{unsupported}),
            };

            return this.writeFailingError(buf, 1);
        },
    } orelse {
        return this.writeFailingError(Builtin.Kind.mkdir.usageString(), 1);
    };

    this.state = .{
        .exec = .{
            .args = filepath_args,
        },
    };

    return this.next();
}

pub fn next(this: *Mkdir) Yield {
    switch (this.state) {
        .idle => @panic("Invalid state"),
        .exec => {
            var exec = &this.state.exec;
            if (exec.started) {
                if (this.state.exec.tasks_done >= this.state.exec.tasks_count and this.state.exec.output_done >= this.state.exec.output_waiting) {
                    const exit_code: ExitCode = if (this.state.exec.err != null) 1 else 0;
                    if (this.state.exec.err) |e| e.deref();
                    this.state = .done;
                    return this.bltn().done(exit_code);
                }
                return .suspended;
            }

            exec.started = true;
            exec.tasks_count = exec.args.len;

            for (exec.args) |dir_to_mk_| {
                const dir_to_mk = dir_to_mk_[0..std.mem.len(dir_to_mk_) :0];
                var task = ShellMkdirTask.create(this, this.opts, dir_to_mk, this.bltn().parentCmd().base.shell.cwdZ());
                task.schedule();
            }
            return .suspended;
        },
        .waiting_write_err => return .failed,
        .done => return this.bltn().done(0),
    }
}

pub fn onShellMkdirTaskDone(this: *Mkdir, task: *ShellMkdirTask) void {
    defer task.deinit();
    this.state.exec.tasks_done += 1;
    var output = task.takeOutput();
    const err = task.err;
    const output_task: *ShellMkdirOutputTask = bun.new(ShellMkdirOutputTask, .{
        .parent = this,
        .output = .{ .arrlist = output.moveToUnmanaged() },
        .state = .waiting_write_err,
    });

    if (err) |e| {
        const error_string = this.bltn().taskErrorToString(.mkdir, e);
        this.state.exec.err = e;
        output_task.start(error_string).run();
        return;
    }
    output_task.start(null).run();
}

pub const ShellMkdirOutputTask = OutputTask(Mkdir, .{
    .writeErr = ShellMkdirOutputTaskVTable.writeErr,
    .onWriteErr = ShellMkdirOutputTaskVTable.onWriteErr,
    .writeOut = ShellMkdirOutputTaskVTable.writeOut,
    .onWriteOut = ShellMkdirOutputTaskVTable.onWriteOut,
    .onDone = ShellMkdirOutputTaskVTable.onDone,
});

const ShellMkdirOutputTaskVTable = struct {
    pub fn writeErr(this: *Mkdir, childptr: anytype, errbuf: []const u8) ?Yield {
        if (this.bltn().stderr.needsIO()) |safeguard| {
            this.state.exec.output_waiting += 1;
            return this.bltn().stderr.enqueue(childptr, errbuf, safeguard);
        }
        _ = this.bltn().writeNoIO(.stderr, errbuf);
        return null;
    }

    pub fn onWriteErr(this: *Mkdir) void {
        this.state.exec.output_done += 1;
    }

    pub fn writeOut(this: *Mkdir, childptr: anytype, output: *OutputSrc) ?Yield {
        if (this.bltn().stdout.needsIO()) |safeguard| {
            this.state.exec.output_waiting += 1;
            const slice = output.slice();
            log("THE SLICE: {d} {s}", .{ slice.len, slice });
            return this.bltn().stdout.enqueue(childptr, slice, safeguard);
        }
        _ = this.bltn().writeNoIO(.stdout, output.slice());
        return null;
    }

    pub fn onWriteOut(this: *Mkdir) void {
        this.state.exec.output_done += 1;
    }

    pub fn onDone(this: *Mkdir) Yield {
        return this.next();
    }
};

pub fn deinit(this: *Mkdir) void {
    _ = this;
}

pub const ShellMkdirTask = struct {
    mkdir: *Mkdir,

    opts: Opts,
    filepath: [:0]const u8,
    cwd_path: [:0]const u8,
    created_directories: ArrayList(u8),

    err: ?jsc.SystemError = null,
    task: jsc.WorkPoolTask = .{ .callback = &runFromThreadPool },
    event_loop: jsc.EventLoopHandle,
    concurrent_task: jsc.EventLoopTask,

    pub fn deinit(this: *ShellMkdirTask) void {
        this.created_directories.deinit();
        bun.default_allocator.destroy(this);
    }

    fn takeOutput(this: *ShellMkdirTask) ArrayList(u8) {
        const out = this.created_directories;
        this.created_directories = ArrayList(u8).init(bun.default_allocator);
        return out;
    }

    pub fn format(this: *const ShellMkdirTask, writer: *std.Io.Writer) !void {
        try writer.print("ShellMkdirTask(0x{x}, filepath={s})", .{ @intFromPtr(this), this.filepath });
    }

    pub fn create(
        mkdir: *Mkdir,
        opts: Opts,
        filepath: [:0]const u8,
        cwd_path: [:0]const u8,
    ) *ShellMkdirTask {
        const task = bun.handleOom(bun.default_allocator.create(ShellMkdirTask));
        const evtloop = mkdir.bltn().parentCmd().base.eventLoop();
        task.* = ShellMkdirTask{
            .mkdir = mkdir,
            .opts = opts,
            .cwd_path = cwd_path,
            .filepath = filepath,
            .created_directories = ArrayList(u8).init(bun.default_allocator),
            .event_loop = evtloop,
            .concurrent_task = jsc.EventLoopTask.fromEventLoop(evtloop),
        };
        return task;
    }

    pub fn schedule(this: *@This()) void {
        debug("{f} schedule", .{this});
        WorkPool.schedule(&this.task);
    }

    pub fn runFromMainThread(this: *@This()) void {
        debug("{f} runFromJS", .{this});
        this.mkdir.onShellMkdirTaskDone(this);
    }

    pub fn runFromMainThreadMini(this: *@This(), _: *void) void {
        this.runFromMainThread();
    }

    fn runFromThreadPool(task: *jsc.WorkPoolTask) void {
        var this: *ShellMkdirTask = @fieldParentPtr("task", task);
        debug("{f} runFromThreadPool", .{this});

        // We have to give an absolute path to our mkdir
        // implementation for it to work with cwd
        const filepath: [:0]const u8 = brk: {
            if (ResolvePath.Platform.auto.isAbsolute(this.filepath)) break :brk this.filepath;
            const parts: []const []const u8 = &.{
                this.cwd_path[0..],
                this.filepath[0..],
            };
            break :brk ResolvePath.joinZ(parts, .auto);
        };

        var node_fs = jsc.Node.fs.NodeFS{};
        // Recursive
        if (this.opts.parents) {
            const args = jsc.Node.fs.Arguments.Mkdir{
                .path = jsc.Node.PathLike{ .string = bun.PathString.init(filepath) },
                .recursive = true,
                .always_return_none = true,
            };

            var vtable = MkdirVerboseVTable{ .inner = this, .active = this.opts.verbose };

            switch (node_fs.mkdirRecursiveImpl(args, *MkdirVerboseVTable, &vtable)) {
                .result => {},
                .err => |e| {
                    this.err = e.withPath(bun.handleOom(bun.default_allocator.dupe(u8, filepath))).toShellSystemError();
                    std.mem.doNotOptimizeAway(&node_fs);
                },
            }
        } else {
            const args = jsc.Node.fs.Arguments.Mkdir{
                .path = jsc.Node.PathLike{ .string = bun.PathString.init(filepath) },
                .recursive = false,
                .always_return_none = true,
            };
            switch (node_fs.mkdirNonRecursive(args)) {
                .result => {
                    if (this.opts.verbose) {
                        bun.handleOom(this.created_directories.appendSlice(filepath[0..filepath.len]));
                        bun.handleOom(this.created_directories.append('\n'));
                    }
                },
                .err => |e| {
                    this.err = e.withPath(bun.handleOom(bun.default_allocator.dupe(u8, filepath))).toShellSystemError();
                    std.mem.doNotOptimizeAway(&node_fs);
                },
            }
        }

        if (this.event_loop == .js) {
            this.event_loop.js.enqueueTaskConcurrent(this.concurrent_task.js.from(this, .manual_deinit));
        } else {
            this.event_loop.mini.enqueueTaskConcurrent(this.concurrent_task.mini.from(this, "runFromMainThreadMini"));
        }
    }

    const MkdirVerboseVTable = struct {
        inner: *ShellMkdirTask,
        active: bool,

        pub fn onCreateDir(vtable: *@This(), dirpath: bun.OSPathSliceZ) void {
            if (!vtable.active) return;
            if (bun.Environment.isWindows) {
                var buf: bun.PathBuffer = undefined;
                const str = bun.strings.fromWPath(&buf, dirpath[0..dirpath.len]);
                bun.handleOom(vtable.inner.created_directories.appendSlice(str));
                bun.handleOom(vtable.inner.created_directories.append('\n'));
            } else {
                bun.handleOom(vtable.inner.created_directories.appendSlice(dirpath));
                bun.handleOom(vtable.inner.created_directories.append('\n'));
            }
            return;
        }
    };
};

const Opts = struct {
    /// -m, --mode
    ///
    /// set file mode (as in chmod), not a=rwx - umask
    mode: ?u32 = null,

    /// -p, --parents
    ///
    /// no error if existing, make parent directories as needed,
    /// with their file modes unaffected by any -m option.
    parents: bool = false,

    /// -v, --verbose
    ///
    /// print a message for each created directory
    verbose: bool = false,

    const Parse = FlagParser(*@This());

    pub fn parse(opts: *Opts, args: []const [*:0]const u8) Result(?[]const [*:0]const u8, ParseError) {
        return Parse.parseFlags(opts, args);
    }

    pub fn parseLong(this: *Opts, flag: []const u8) ?ParseFlagResult {
        if (bun.strings.eqlComptime(flag, "--mode")) {
            return .{ .unsupported = "--mode" };
        } else if (bun.strings.eqlComptime(flag, "--parents")) {
            this.parents = true;
            return .continue_parsing;
        } else if (bun.strings.eqlComptime(flag, "--vebose")) {
            this.verbose = true;
            return .continue_parsing;
        }

        return null;
    }

    pub fn parseShort(this: *Opts, char: u8, smallflags: []const u8, i: usize) ?ParseFlagResult {
        switch (char) {
            'm' => {
                return .{ .unsupported = "-m " };
            },
            'p' => {
                this.parents = true;
            },
            'v' => {
                this.verbose = true;
            },
            else => {
                return .{ .illegal_option = smallflags[1 + i ..] };
            },
        }

        return null;
    }
};

pub inline fn bltn(this: *Mkdir) *Builtin {
    const impl: *Builtin.Impl = @alignCast(@fieldParentPtr("mkdir", this));
    return @fieldParentPtr("impl", impl);
}

// --
const debug = bun.Output.scoped(.ShellMkdir, .hidden);

const log = debug;

const std = @import("std");
const ArrayList = std.array_list.Managed;

const interpreter = @import("../interpreter.zig");
const FlagParser = interpreter.FlagParser;
const Interpreter = interpreter.Interpreter;
const OutputSrc = interpreter.OutputSrc;
const OutputTask = interpreter.OutputTask;
const ParseError = interpreter.ParseError;
const ParseFlagResult = interpreter.ParseFlagResult;

const Builtin = Interpreter.Builtin;
const Result = Interpreter.Builtin.Result;

const bun = @import("bun");
const ResolvePath = bun.path;

const jsc = bun.jsc;
const WorkPool = bun.jsc.WorkPool;

const shell = bun.shell;
const ExitCode = shell.ExitCode;
const Yield = bun.shell.Yield;

File: src/runtime/shell/builtin/mkdir.rs
use crate::node::fs::{MkdirCtx, NodeFS, args as fs_args};
use crate::node::types::PathLike;
use crate::shell::ExitCode;
use crate::shell::builtin::{Builtin, BuiltinState, IoKind, Kind};
use crate::shell::interpreter::{
    FlagParser, Interpreter, NodeId, OutputSrc, OutputTask, OutputTaskVTable, OutputWrite,
    ParseFlagResult, ShellTask, unsupported_flag,
};
use crate::shell::io_writer::{ChildPtr, WriterTag};
use crate::shell::yield_::Yield;

#[derive(Default)]
pub struct Mkdir {
    pub(crate) opts: Opts,
    pub(crate) state: State,
}

#[derive(Default)]
pub enum State {
    #[default]
    Idle,
    Exec(Exec),
    WaitingWriteErr,
    Done,
}

pub struct Exec {
    pub(crate) started: bool,
    pub(crate) tasks_count: usize,
    pub(crate) tasks_done: usize,
    pub(crate) output_waiting: u16,
    pub(crate) output_done: u16,
    /// Index into `Builtin::args` where filepath args start (storing the
    /// index keeps the lifetime tied to the Cmd's argv without a
    /// self-reference).
    pub(crate) args_start: usize,
    pub(crate) err: Option<bun_sys::Error>,
    /// FIFO of in-flight OutputTasks awaiting an IOWriter chunk completion
    /// (`WriterTag` cannot name an OutputTask): `write_err`/`write_out`
    /// push, `on_io_writer_chunk` pops and forwards to
    /// `OutputTask::on_io_writer_chunk` so the writeErr→writeOut→onDone
    /// state machine runs.
    pub(crate) output_queue: std::collections::VecDeque<Box<OutputTask<Mkdir>>>,
}

impl Mkdir {
    pub(crate) fn start(interp: &Interpreter, cmd: NodeId) -> Yield {
        let (args_start, mut opts) = {
            let mut opts = Opts::default();
            match Builtin::parse_flags(interp, cmd, &mut opts) {
                Ok(Some(start)) => (start, opts),
                Ok(None) => {
                    return Self::fail_usage(interp, cmd);
                }
                Err(e) => {
                    return Builtin::fail_parse(interp, cmd, Kind::Mkdir, &e, || {
                        Self::state_mut(interp, cmd).state = State::WaitingWriteErr
                    });
                }
            }
        };
        // Hand the parsed opts back into state.
        core::mem::swap(&mut Self::state_mut(interp, cmd).opts, &mut opts);

        Self::state_mut(interp, cmd).state = State::Exec(Exec {
            started: false,
            tasks_count: 0,
            tasks_done: 0,
            output_waiting: 0,
            output_done: 0,
            args_start,
            err: None,
            output_queue: std::collections::VecDeque::new(),
        });
        Self::next(interp, cmd)
    }

    fn fail_usage(interp: &Interpreter, cmd: NodeId) -> Yield {
        Self::state_mut(interp, cmd).state = State::WaitingWriteErr;
        Builtin::write_failing_error(interp, cmd, Kind::Mkdir.usage_string(), 1)
    }

    fn next(interp: &Interpreter, cmd: NodeId) -> Yield {
        // Read scalars, drop the borrow, then act.
        let action = match &mut Self::state_mut(interp, cmd).state {
            State::Idle => panic!("Invalid state"),
            State::Exec(exec) => {
                if exec.started {
                    if exec.tasks_done >= exec.tasks_count
                        && exec.output_done >= exec.output_waiting
                    {
                        let exit_code: ExitCode = if exec.err.is_some() { 1 } else { 0 };
                        exec.err = None;
                        NextAction::Done(exit_code)
                    } else {
                        NextAction::Suspend
                    }
                } else {
                    exec.started = true;
                    NextAction::Schedule(exec.args_start)
                }
            }
            State::WaitingWriteErr => NextAction::Failed,
            State::Done => NextAction::AlreadyDone,
        };
        match action {
            NextAction::Suspend => Yield::suspended(),
            NextAction::Failed => Yield::failed(),
            NextAction::AlreadyDone => Builtin::done(interp, cmd, 0),
            NextAction::Done(code) => {
                Self::state_mut(interp, cmd).state = State::Done;
                Builtin::done(interp, cmd, code)
            }
            NextAction::Schedule(args_start) => {
                let argc = Builtin::argc(interp, cmd);
                let task_count = argc - args_start;
                if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                    exec.tasks_count = task_count;
                }
                let opts = Self::state_mut(interp, cmd).opts;
                let cwd = Builtin::shell(interp, cmd).borrow().cwd().to_vec();
                for i in args_start..argc {
                    let path = Builtin::of(interp, cmd).arg_bytes(i).to_vec();
                    let task = ShellMkdirTask::create(cmd, opts, path, cwd.clone(), interp);
                    ShellTask::schedule(task);
                }
                Yield::suspended()
            }
        }
    }

    pub(crate) fn on_io_writer_chunk(
        interp: &Interpreter,
        cmd: NodeId,
        written: usize,
        e: Option<bun_sys::SystemError>,
    ) -> Yield {
        let pending = match &mut Self::state_mut(interp, cmd).state {
            State::WaitingWriteErr => Err(()),
            State::Exec(exec) => Ok(exec.output_queue.pop_front()),
            State::Idle | State::Done => panic!("Invalid state"),
        };
        let Ok(pending) = pending else {
            return Builtin::done(interp, cmd, 1);
        };
        if let Some(task) = pending {
            return OutputTask::<Mkdir>::on_io_writer_chunk(task, interp, written, e);
        }
        Self::next(interp, cmd)
    }

    /// The caller ([`ShellMkdirTask::run_from_main_thread`]) owns the heap
    /// allocation and drops it after this returns.
    fn on_shell_mkdir_task_done(interp: &Interpreter, cmd: NodeId, task: &mut ShellMkdirTask) {
        let output = core::mem::take(&mut task.created_directories);
        let err = task.err.take();
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.tasks_done += 1;
        }

        let output_task = OutputTask::<Mkdir>::new(cmd, OutputSrc::Arrlist(output));
        let errstr: Option<Vec<u8>> = err.map(|e| {
            let s = Builtin::task_error_to_string(Kind::Mkdir, &e);
            if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                exec.err = Some(e);
            }
            s
        });
        OutputTask::<Mkdir>::start(output_task, interp, errstr.as_deref()).run(interp);
    }
}

enum NextAction {
    Done(ExitCode),
    Schedule(usize),
    Suspend,
    Failed,
    AlreadyDone,
}

impl OutputTaskVTable for Mkdir {
    fn write_err(
        interp: &Interpreter,
        cmd: NodeId,
        child: Box<OutputTask<Self>>,
        errbuf: &[u8],
    ) -> OutputWrite<Self> {
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.output_waiting += 1;
        }
        let stderr_needs_io = Builtin::of(interp, cmd).stderr.needs_io();
        if let Some(safeguard) = stderr_needs_io {
            // OutputTask has no `WriterTag` of its own (it is not directly
            // dispatchable as an IOWriter child), so the enqueue is tagged
            // `WriterTag::Builtin` and `child` is parked on `output_queue`;
            // `on_io_writer_chunk` pops it to route the completion back to
            // the OutputTask state machine.
            if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                exec.output_queue.push_back(child);
            }
            let childptr = ChildPtr::new(cmd, WriterTag::Builtin);
            return OutputWrite::Enqueued(Builtin::write_out(
                interp,
                cmd,
                IoKind::Stderr,
                childptr,
                errbuf,
                safeguard,
            ));
        }
        let _ = Builtin::write_no_io(interp, cmd, IoKind::Stderr, errbuf);
        OutputWrite::Done(child)
    }

    fn on_write_err(interp: &Interpreter, cmd: NodeId) {
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.output_done += 1;
        }
    }

    fn write_out(
        interp: &Interpreter,
        cmd: NodeId,
        child: Box<OutputTask<Self>>,
    ) -> OutputWrite<Self> {
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.output_waiting += 1;
        }
        let stdout_needs_io = Builtin::of(interp, cmd).stdout.needs_io();
        if let Some(safeguard) = stdout_needs_io {
            // See write_err — park `child` so the chunk callback routes to
            // OutputTask::on_io_writer_chunk.
            let childptr = ChildPtr::new(cmd, WriterTag::Builtin);
            return OutputWrite::Enqueued(Builtin::write_out_with(
                interp,
                cmd,
                IoKind::Stdout,
                childptr,
                safeguard,
                |buf| {
                    buf.extend_from_slice(child.output.slice());
                    if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
                        exec.output_queue.push_back(child);
                    }
                },
            ));
        }
        let _ = Builtin::write_no_io(interp, cmd, IoKind::Stdout, child.output.slice());
        OutputWrite::Done(child)
    }

    fn on_write_out(interp: &Interpreter, cmd: NodeId) {
        if let State::Exec(exec) = &mut Self::state_mut(interp, cmd).state {
            exec.output_done += 1;
        }
    }

    fn on_done(interp: &Interpreter, cmd: NodeId) -> Yield {
        Self::next(interp, cmd)
    }
}

/// Runs `mkdir`/`mkdir -p` on a worker
/// thread, then bounces back to the main thread.
pub(crate) struct ShellMkdirTask {
    /// Owning Cmd node (the mkdir builtin's id).
    pub cmd: NodeId,
    pub opts: Opts,
    /// Owned copy of the target path (owned to avoid threading a lifetime
    /// through the WorkPool).
    pub filepath: Vec<u8>,
    pub cwd_path: Vec<u8>,
    pub created_directories: Vec<u8>,
    pub err: Option<bun_sys::Error>,
    pub task: ShellTask,
}

crate::shell_task!(ShellMkdirTask);

impl ShellMkdirTask {
    fn create(
        cmd: NodeId,
        opts: Opts,
        filepath: Vec<u8>,
        cwd_path: Vec<u8>,
        interp: &Interpreter,
    ) -> Box<ShellMkdirTask> {
        Box::new(ShellMkdirTask {
            cmd,
            opts,
            filepath,
            cwd_path,
            created_directories: Vec::new(),
            err: None,
            task: ShellTask::new(interp),
        })
    }

    fn run_from_thread_pool(this: &mut ShellMkdirTask) {
        use bun_paths::{Platform, platform, resolve_path};
        // We have to give an absolute path to our mkdir implementation for it
        // to work with cwd.
        let mut spill = Vec::new();
        let filepath: &bun_core::ZStr = if Platform::AUTO.is_absolute(&this.filepath) {
            // Owned `Vec<u8>`; ensure NUL-terminated.
            if this.filepath.last() != Some(&0) {
                this.filepath.push(0);
            }
            bun_core::ZStr::from_buf(&this.filepath, this.filepath.len() - 1)
        } else {
            resolve_path::join_z_spill::<platform::Auto>(
                &mut spill,
                &[&this.cwd_path, &this.filepath],
            )
        };

        // `NodeFS` expects the `Valid::path_too_long` bound its JS callers
        // enforce; past it, `PathLike::slice_z` yields "" and mkdir reports ENOENT.
        if filepath.len() >= bun_paths::MAX_PATH_BYTES {
            this.err = Some(
                bun_sys::Error::from_code(bun_sys::E::ENAMETOOLONG, bun_sys::Tag::mkdir)
                    .with_path(filepath.as_bytes()),
            );
            return;
        }

        let mut node_fs = NodeFS::default();
        let args = fs_args::Mkdir {
            path: PathLike::borrowed(filepath.as_bytes()),
            recursive: this.opts.parents,
            mode: fs_args::Mkdir::DEFAULT_MODE,
            always_return_none: true,
        };

        if this.opts.parents {
            let vtable = MkdirVerboseVTable {
                inner: &raw mut this.created_directories,
                active: this.opts.verbose,
            };
            if let Err(e) = node_fs.mkdir_recursive_impl(&args, &vtable) {
                this.err = Some(e.with_path(filepath.as_bytes()));
                core::hint::black_box(&node_fs);
            }
        } else {
            match node_fs.mkdir_non_recursive(&args) {
                Ok(_) => {
                    if this.opts.verbose {
                        this.created_directories
                            .extend_from_slice(filepath.as_bytes());
                        this.created_directories.push(b'\n');
                    }
                }
                Err(e) => {
                    this.err = Some(e.with_path(filepath.as_bytes()));
                    core::hint::black_box(&node_fs);
                }
            }
        }
        // Bounce-back to the main thread is posted by `ShellTask::run_owned`
        // via `ShellTask::on_finish` (handles both JS and mini event loops).
    }
}

// `runtime::dispatch::run_task`'s `task_tag::ShellMkdirTask` arm reboxes the
// pointer `ShellTask::on_finish` posted; a completion that will not run drops
// the keep-alive and the box.

/// Collects each created directory into
/// `created_directories` (newline-separated) when `-v` is set. Passed by value
/// to `NodeFS::mkdir_recursive_impl`; `on_create_dir` writes through the raw
/// back-ref because the trait method takes `&self`.
struct MkdirVerboseVTable {
    inner: *mut Vec<u8>,
    active: bool,
}

impl MkdirCtx for MkdirVerboseVTable {
    fn on_create_dir(&self, dirpath: &bun_paths::OSPathSliceZ) {
        if !self.active {
            return;
        }
        // SAFETY: `inner` points at `ShellMkdirTask::created_directories`; the
        // worker thread is the sole accessor for the duration of
        // `run_from_thread_pool`, and `mkdir_recursive_impl` does not alias it.
        let out = unsafe { &mut *self.inner };
        #[cfg(windows)]
        {
            let mut buf = bun_paths::PathBuffer::uninit();
            let str = bun_paths::strings::from_wpath(buf.as_mut(), dirpath.as_slice());
            out.extend_from_slice(str.as_bytes());
            out.push(b'\n');
        }
        #[cfg(not(windows))]
        {
            out.extend_from_slice(dirpath.as_bytes());
            out.push(b'\n');
        }
    }
}

impl crate::shell::interpreter::ShellTaskCtx for ShellMkdirTask {
    fn shell_task(&self) -> &ShellTask {
        &self.task
    }
    fn shell_task_mut(&mut self) -> &mut ShellTask {
        &mut self.task
    }
    fn run_from_thread_pool(&mut self) {
        Self::run_from_thread_pool(self)
    }
    /// The task drops after `on_shell_mkdir_task_done` has taken what it needs.
    fn run_from_main_thread(mut self: Box<Self>, interp: &Interpreter) {
        let cmd = self.cmd;
        Mkdir::on_shell_mkdir_task_done(interp, cmd, &mut self);
    }
}

#[derive(Default, Clone, Copy)]
pub struct Opts {
    /// `-p`, `--parents` — no error if existing, make parent directories as
    /// needed, with their file modes unaffected by any -m option.
    pub(crate) parents: bool,
    /// `-v`, `--verbose` — print a message for each created directory
    pub(crate) verbose: bool,
}

impl FlagParser for Opts {
    fn parse_long(&mut self, flag: &[u8]) -> Option<ParseFlagResult> {
        if flag == b"--mode" {
            return Some(ParseFlagResult::Unsupported(unsupported_flag(b"--mode")));
        }
        if flag == b"--parents" {
            self.parents = true;
            return Some(ParseFlagResult::ContinueParsing);
        }
        // Note: the `--vebose` typo is intentional (kept for compatibility).
        if flag == b"--vebose" {
            self.verbose = true;
            return Some(ParseFlagResult::ContinueParsing);
        }
        None
    }

    fn parse_short(&mut self, ch: u8, smallflags: &[u8], i: usize) -> Option<ParseFlagResult> {
        match ch {
            b'm' => Some(ParseFlagResult::Unsupported(unsupported_flag(b"-m "))),
            b'p' => {
                self.parents = true;
                None
            }
            b'v' => {
                self.verbose = true;
                None
            }
            _ => Some(ParseFlagResult::IllegalOption(smallflags[1 + i..].into())),
        }
    }
}
